examples/webview2/src/main.rs

```diff
@@ -0,0 +1,622 @@
+    fn drop(&mut self) {
+        match Arc::strong_count(&self.controller) {
+            0 => {
+                self.controller.Close().expect("call Close");
+            }
+            _ => (),
+        }
+    }
+}
+
+#[derive(Debug, Deserialize)]
+struct InvokeMessage {
+    id: u64,
+    method: String,
+    params: Vec<Value>,
+}
+
+impl WebView {
+    pub fn create(debug: bool, parent: Option<HWND>) -> Result<WebView> {
+        let (parent, frame) = match parent {
+            Some(hwnd) => (hwnd, None),
+            None => {
+                let frame = FrameWindow::new();
+                (*frame.window, Some(frame))
+            }
+        };
+
+        let (tx, rx) = mpsc::channel();
+        let environment = {
+            CoreWebView2Environment::CreateAsync()?.SetCompleted(
+                AsyncOperationCompletedHandler::new(move |op, _status| {
+                    if let Some(op) = op {
+                        tx.send(op.GetResults()?).expect("send over mpsc channel");
+                    }
+                    Ok(())
+                }),
+            )?;
+
+            wait_with_pump(rx)
+        }?;
+
+        let (tx, rx) = mpsc::channel();
+        let controller = {
+            environment
+                .CreateCoreWebView2ControllerAsync(
+                    CoreWebView2ControllerWindowReference::CreateFromWindowHandle(parent.0 as u64)?,
+                )?
+                .SetCompleted(AsyncOperationCompletedHandler::new(move |op, _status| {
+                    if let Some(op) = op.as_ref() {
+                        tx.send(op.GetResults()?).expect("send over mpsc channel");
+                    }
+                    Ok(())
+                }))?;
+
+            wait_with_pump(rx)
+        }?;
+
+        let size = get_window_size(parent);
+        let mut client_rect = RECT::default();
+        unsafe { WindowsAndMessaging::GetClientRect(parent, &mut client_rect) };
+        controller.SetBounds(Rect {
+            X: 0f32,
+            Y: 0f32,
+            Width: size.cx as f32,
+            Height: size.cy as f32,
+        })?;
+        controller.SetIsVisible(true)?;
+
+        let webview = controller.CoreWebView2()?;
+
+        if !debug {
+            let settings = webview.Settings()?;
+            settings.SetAreDevToolsEnabled(false)?;
+            settings.SetAreDefaultContextMenusEnabled(false)?;
+        }
+
+        let bindings: Arc<Mutex<HashMap<String, Box<dyn FnMut(u64, Vec<Value>)>>>> =
+            Arc::new(Mutex::new(HashMap::new()));
+        let bindings_ref = bindings.clone();
+        let token = webview.WebMessageReceived(TypedEventHandler::<
+            CoreWebView2,
+            CoreWebView2WebMessageReceivedEventArgs,
+        >::new(move |_sender, args| {
+            if let Some(args) = args {
+                if let Ok(message) = String::from_utf16(args.WebMessageAsJson()?.as_wide()) {
+                    if let Ok(value) = serde_json::from_str::<InvokeMessage>(&message) {
```

Comment:
I suppose I could use `Windows::Data::Json` for that, but I think `serde_json` is enough of a de-facto standard for Rust that it's more valuable to show the interop. I think I'll leave this one as is unless you feel strongly that it should use the Windows API.

Reply:
My view is to use whatever is simplest. The challenge with examples like this is to keep them being good examples of doing X - where X is singular - rather than becoming a tour de force of all the things. 😉 It's a fine line - just something to keep in mind.